### PR TITLE
NIAD-NA-security-scans-in-pipeline

### PR DESCRIPTION
**Pull request by michalschott**
Small stages refactor in Jenkinsfile.
Added trivy as a security scanner in pipeline (will fail if any fixable vulnerabilities are present).

Fixes:
```
+---------------+------------------+----------+-------------------+-------------------+--------------------------------+------------------------------------+
 |    LIBRARY    | VULNERABILITY ID | SEVERITY | INSTALLED VERSION |   FIXED VERSION   |             TITLE              |                URL                 |
 +---------------+------------------+----------+-------------------+-------------------+--------------------------------+------------------------------------+
 | apt           | CVE-2020-27350   | MEDIUM   | 2.0.2ubuntu0.1    | 2.0.2ubuntu0.2    | APT had several integer        | avd.aquasec.com/nvd/cve-2020-27350 |
 |               |                  |          |                   |                   | overflows and underflows while |                                    |
 |               |                  |          |                   |                   | parsing .deb packages, aka...  |                                    |
 +---------------+------------------+          +-------------------+-------------------+--------------------------------+------------------------------------+
 | curl          | CVE-2020-8285    |          | 7.68.0-1ubuntu2.2 | 7.68.0-1ubuntu2.4 | curl: malicious FTP server can | avd.aquasec.com/nvd/cve-2020-8285  |
 |               |                  |          |                   |                   | trigger stack overflow when    |                                    |
 |               |                  |          |                   |                   | CURLOPT_CHUNK_BGN_FUNCTION is  |                                    |
 |               |                  |          |                   |                   | used...                        |                                    |
 +               +------------------+          +                   +                   +--------------------------------+------------------------------------+
 |               | CVE-2020-8286    |          |                   |                   | curl: inferior OCSP            | avd.aquasec.com/nvd/cve-2020-8286  |
 |               |                  |          |                   |                   | verification                   |                                    |
 +               +------------------+----------+                   +                   +--------------------------------+------------------------------------+
 |               | CVE-2020-8284    | LOW      |                   |                   | curl: dangerous nature of PASV | avd.aquasec.com/nvd/cve-2020-8284  |
 |               |                  |          |                   |                   | command could be used to make  |                                    |
 |               |                  |          |                   |                   | curl...                        |                                    |
 +---------------+------------------+----------+-------------------+-------------------+--------------------------------+------------------------------------+
 | libapt-pkg6.0 | CVE-2020-27350   | MEDIUM   | 2.0.2ubuntu0.1    | 2.0.2ubuntu0.2    | APT had several integer        | avd.aquasec.com/nvd/cve-2020-27350 |
 |               |                  |          |                   |                   | overflows and underflows while |                                    |
 |               |                  |          |                   |                   | parsing .deb packages, aka...  |                                    |
 +---------------+------------------+          +-------------------+-------------------+--------------------------------+------------------------------------+
 | libcurl4      | CVE-2020-8285    |          | 7.68.0-1ubuntu2.2 | 7.68.0-1ubuntu2.4 | curl: malicious FTP server can | avd.aquasec.com/nvd/cve-2020-8285  |
 |               |                  |          |                   |                   | trigger stack overflow when    |                                    |
 |               |                  |          |                   |                   | CURLOPT_CHUNK_BGN_FUNCTION is  |                                    |
 |               |                  |          |                   |                   | used...                        |                                    |
 +               +------------------+          +                   +                   +--------------------------------+------------------------------------+
 |               | CVE-2020-8286    |          |                   |                   | curl: inferior OCSP            | avd.aquasec.com/nvd/cve-2020-8286  |
 |               |                  |          |                   |                   | verification                   |                                    |
 +               +------------------+----------+                   +                   +--------------------------------+------------------------------------+
 |               | CVE-2020-8284    | LOW      |                   |                   | curl: dangerous nature of PASV | avd.aquasec.com/nvd/cve-2020-8284  |
 |               |                  |          |                   |                   | command could be used to make  |                                    |
 |               |                  |          |                   |                   | curl...                        |                                    |
 +---------------+------------------+----------+-------------------+-------------------+--------------------------------+------------------------------------+
 | libssl1.1     | CVE-2020-1971    | HIGH     | 1.1.1f-1ubuntu2   | 1.1.1f-1ubuntu2.1 | openssl: EDIPARTYNAME NULL     | avd.aquasec.com/nvd/cve-2020-1971  |
 |               |                  |          |                   |                   | pointer de-reference           |                                    |
 +---------------+                  +          +                   +                   +                                +                                    +
 | openssl       |                  |          |                   |                   |                                |                                    |
 |               |                  |          |                   |                   |                                |                                    |
 +---------------+------------------+----------+-------------------+-------------------+--------------------------------+------------------------------------+
```

### CODE REVIEW

File: Jenkinsfile [hunk 1]
```diff
@@ -15,62 +15,71 @@ pipeline {
...
+                always {
+                    step([
+                        $class : 'JacocoPublisher',
+                        execPattern : '**/build/jacoco/*.exec',
+                        classPattern : '**/build/classes/java',
+                        sourcePattern : 'src/main/java',
+                        exclusionPattern : '**/*Test.class'
+                    ])
+            sh "rm -rf build"
+                    sh "docker-compose -f docker/docker-compose.yml -f docker/docker-compose-tests.yml down"
```
Comment (bartek-sarul) [inline Jenkinsfile:48]: It is worth to dump all container logs into artefacts

**Review by bartek-sarul — CHANGES_REQUESTED** [review]

File: Jenkinsfile (hunk 1 above)
Comment (michalschott) [reply Jenkinsfile:48]: main purpose of this PR is to:
* add stage where we will do vulnerability scanning and fail pipeline if there are any which can be fixed
* fix every vulnerability in container which is fixable
* remove blank/unused "Build" stage

feel free to raise another PR with your proposal @bartek-sarul

**Comment by mattd-kainos** [comment]
Vulnerability scans should not fail development pipeline builds. 1) Sometime the vulnerability cannot be fixed and this blocked development. 2) It forces developers to include unrelated work (upgrading dependencies to fix CVEs) as part of their PRs

These types of scans belong in a separate pipeline and any fixes are part of planned sprint work for visibility.

**Comment by michalschott** [comment]
> Vulnerability scans should not fail development pipeline builds. 1) Sometime the vulnerability cannot be fixed and this blocked development. 2) It forces developers to include unrelated work (upgrading dependencies to fix CVEs) as part of their PRs
> 
> These types of scans belong in a separate pipeline and any fixes are part of planned sprint work for visibility.

`--ignore-unfixed` is included to not fail pipeline

**Review by mattd-kainos — CHANGES_REQUESTED** [review]
CVE scans of third party dependencies should not be part of the development pipeline

**Comment by michalschott** [comment]
> CVE scans of third party dependencies should not be part of the development pipeline

I strongly disagree, someone could potentially introduce serious backdoors (even unintentionally) without even anyone noting this.

**Comment by michalschott** [comment]
@petervdm can you review please

**Comment by mattd-kainos** [comment]
snyk and docker scan are used during the release process

**Pull request closed without merging**